exporter/stefexporter/exporter.go

```diff
@@ -5,26 +5,395 @@ package stefexporter // import "github.com/open-telemetry/opentelemetry-collecto
+	stefWriter      *oteltef.MetricsWriter
+	stefWriterMutex sync.Mutex // protects stefWriter
+
+	// Ack-related fields follow and are protected by ackDataMutex.
+	ackDataMutex sync.Mutex
+	// Last ack ID received from the server.
+	lastAckedID uint64
+	// Flag to set to interrupt all waiting for acks. This is set when disconnecting.
+	interruptAckWaits bool
+	// A slice of deadlines waiting for acks.
+	ackExpectedBy []time.Time
+	// Condition used to notify ack waiters. Works in pair with ackDataMutex.
+	ackCond *sync.Cond
+
+	// Indicates ack wait operations that are ongoing.
+	inProgressWaits sync.WaitGroup
+	// Timer that fires when ack waiting time exceeds maxAckWaitTime.
+	ackTimer *time.Timer
+
+	// Channel to stop ticker() goroutine.
+	stopTicker chan struct{}
+}
+
+type loggerWrapper struct {
+	logger *zap.Logger
+}
+
+func (w *loggerWrapper) Debugf(_ context.Context, format string, v ...any) {
+	w.logger.Debug(fmt.Sprintf(format, v...))
 }
 
-func (s *stefExporter) Start(_ context.Context, _ component.Host) error {
+func (w *loggerWrapper) Errorf(_ context.Context, format string, v ...any) {
+	w.logger.Error(fmt.Sprintf(format, v...))
+}
+
+// By default, wait up to 30 seconds for acks to arrive from the server.
+const defMaxAckWaitTime = 30 * time.Second
+
+var errAckTimeout = errors.New("ack timed out")
+
+func newStefExporter(set component.TelemetrySettings, cfg *Config) *stefExporter {
+	exp := &stefExporter{
+		set:                 set,
+		logger:              set.Logger,
+		cfg:                 cfg,
+		maxWritesInProgress: int64(cfg.NumConsumers),
+		ackTimer:            time.NewTimer(time.Duration(1<<63 - 1)),
+		maxAckWaitTime:      cfg.maxAckWaitTime,
+	}
+
+	exp.ackCond = sync.NewCond(&exp.ackDataMutex)
+	if exp.maxAckWaitTime == 0 {
+		exp.maxAckWaitTime = defMaxAckWaitTime
+	}
+
+	exp.compression = pkg.CompressionNone
+	if cfg.Compression == "zstd" {
+		exp.compression = pkg.CompressionZstd
+	}
+	return exp
+}
+
+func (s *stefExporter) Start(ctx context.Context, host component.Host) error {
+	s.host = host
+
+	// No need to block Start(), we will begin connection attempt in a goroutine.
+	go func() {
+		if err := s.ensureConnected(ctx, host); err != nil {
+			s.logger.Error("Error connecting to destination", zap.Error(err))
+			// exportMetrics() will try to connect again as needed.
+		}
+	}()
 	return nil
 }
 
+func (s *stefExporter) ensureConnected(ctx context.Context, host component.Host) error {
+	s.connMutex.Lock()
+	defer s.connMutex.Unlock()
+
+	if s.isConnected {
+		return nil
+	}
+
+	s.logger.Debug("Connecting to destination", zap.String("endpoint", s.cfg.Endpoint))
+
+	// Initialize ack-related fields.
+	s.lastAckedID = 0
+	s.interruptAckWaits = false
+	s.ackExpectedBy = s.ackExpectedBy[:0]
+
+	// Connect to the server.
+	var err error
+	s.grpcConn, err = s.cfg.ClientConfig.ToClientConn(ctx, host, s.set)
+	if err != nil {
+		return err
+	}
```

Comment:
You should not re-create the connection. gRPC is smart enough to re-connect itself, I am confused and never seen this before, which makes me think this is not ideal.

Reply:
Fixed.